Выполнил ДЗ4

bilinin opened the pull request.

bilinin committed:
Добавил файл main.py

bilinin committed:
добавлено задание начального состояния льва

bilinin committed:
добавил файл tests.py

bilinin committed:
добавил .travis.yml

bilinin committed:
Добавил бэйдж

skyline-gleb commented on main.py:
```diff
@@ -0,0 +1,25 @@
+            self.action = ""
+            self.state = state
+        else:
+            raise ValueError("Состояние не определено")
```
хорошая практика заинвертить условие, тогда код который позитивный, пишется ниже, без лишних отступов и читать удобно и все такое

bilinin replied:
поправил

skyline-gleb commented on main.py:
```diff
@@ -0,0 +1,25 @@
+            self.state = lion_get[self.state][symbol]['состояние']
+            return self
+        except KeyError:
+            return self
```
зачем возвращать `self`? ну и вдобавок 2 момента - ошибку бы можно и самому отслеживать что выливается во второе, что ошибку то и более понтяную стоит пользователю вещать. а то пишу код, а как будете пользоваться это ваши проблемы

bilinin replied:
поправил

bilinin committed:
инвертировал условие

bilinin committed:
Добавил более понятное сообщение об ошибке и ValueError()

skyline-gleb commented on main.py:
```diff
@@ -0,0 +1,24 @@
+            raise ValueError("Состояние не определено")
+        else:
+            self.action = ""
+            self.state = state
```
а теперь посмотри на это и поправь. не буду говорить как, чтоб сам осознал)) подсказку, если надо, иди в прошлом комментарий на это место кода.

skyline-gleb commented on main.py:
```diff
@@ -0,0 +1,24 @@
+            raise ValueError('Неверный событие')
+        else:
+            self.action = lion_get[self.state][symbol]['действие']
+            self.state = lion_get[self.state][symbol]['состояние']
```
таж фигня

skyline-gleb commented:
у тебя сейчас входные данные подаются непойми откуда и как, это не правильно. в том числе и тесты делают лишнее и их могло быть и меньше и проще

bilinin committed:
убрал лишние else

bilinin commented:
вот с входными данными не понял, что не так.

skyline-gleb commented:
глобальное объявление таблицы в файле со львом - вот в чем проблема входных данных. убираешь и все легко сломалось. и дело то в том, что убрать её оттуда очень даже логично, она там не нужна.
дело то даже в том, что у меня нет впечатления, что идею таблицы придумал сам. если бы это было так, то ты бы и использовал её по назначению, а сейчас толку от нее не получаешь никакого. и мои намеки на верное использование тебе совсем не подталкивают

bilinin committed:
перенёс таблицу в класс.

bilinin commented:
Ну да, намного логичнее её в класс поместить, затупил.

skyline-gleb commented on main.py:
```diff
@@ -0,0 +1,21 @@
+
+    def __init__(self, state):
+        if state != "голодный" and state != "сытый":
+            raise ValueError("Состояние не определено")
```
тут разве таблица тебе не в помощь?

skyline-gleb commented on main.py:
```diff
@@ -0,0 +1,21 @@
+
+    def input(self, symbol):
+        if symbol != "антилопа" and symbol != "охотник" and symbol != "дерево":
+            raise ValueError('Неверный событие')
```
а здесь?

skyline-gleb commented:
> Ну да, намного логичнее её в класс поместить, затупил.

нет, совсем не логичней. идея таблицы в том чтоб отвязать реализацию конечного автомата и льва, чтоб можно было добавить пару строк в таблицу и все продолжало работать, но у тебя увы. 
SOLID помнишь? И если быть точнее букавка О https://en.wikipedia.org/wiki/Open/closed_principle и я на занятии даже сказал, что большенство реализаций льва не подчинается этому принципу.

bilinin committed:
Теперь функция FSM принимает на вход словарь, по которому работает автомат.

bilinin committed:
Update main.py

bilinin committed:
один тест, который перебирает весь словарь

bilinin commented:
теперь можно и пару строк добавить, а можно и другую таблицу ему дать

bilinin committed:
поправил импорт из main файла

bilinin committed:
переделал структуру словаря

На мой взгляд сделал структуру словаря понятнее, чтобы вытащить состояния теперь dictionary["state"].keys() вместо непонятного dictionary.keys() и т.д.

bilinin committed:
Update tests.py

skyline-gleb commented:
ох вообще не туда ты пошел. формат таблицы усложнил, тест сделала не правильный, так не делается. он и сложный, и куча лишних проверок и т.п. и вообще одним тестом точно не обойтись.
кто мешает спрашивать/задавать вопросы чтоб не делать ерунду?

bilinin commented:
Так, какой тропой мне дальше то идти? Вернуть обратно таблицу, вернуть тесты, а дальше что ?

bilinin committed:
выпилил таблицу

bilinin committed:
Update main.py

bilinin committed:
Update main.py

bilinin committed:
Переместил таблицу сюда, переписал тесты.

skyline-gleb commented on main.py:
```diff
@@ -0,0 +1,15 @@
+    def __init__(self, dictionary, state):
+        if not state in dictionary.keys():
+            raise ValueError("Состояния нет в словаре переходов")
+        else:
```
лишнее же `else`, вообще прям лишнее

skyline-gleb commented on main.py:
```diff
@@ -0,0 +1,15 @@
+    def input(self, symbol):
+        if not symbol in self.dictionary[self.state].keys():
+            raise ValueError('Входного символа нет в словаре переходов')
+        else:
```
таже история

skyline-gleb commented on tests.py:
```diff
@@ -0,0 +1,42 @@
+                     "дерево": {"action": "спать", "state": "голодный"}},
+        "сытый": {"антилопа": {"action": "спать", "state": "голодный"},
+                  "охотник": {"action": "убежать", "state": "голодный"},
+                  "дерево": {"action": "смотреть", "state": "голодный"}}}
```
для тестирования не нужная такая большая таблица

bilinin committed:
убрал лишние else

bilinin committed:
урезал таблицу
